flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/source/reader/HBaseSourceSplitReader.java

```diff
@@ -0,0 +1,131 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.hbase.source.reader;
+
+import org.apache.flink.connector.base.source.reader.RecordsWithSplitIds;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitReader;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsAddition;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsChange;
+import org.apache.flink.connector.hbase.source.hbaseendpoint.HBaseEndpoint;
+import org.apache.flink.connector.hbase.source.split.HBaseSourceSplit;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+
+/** A {@link SplitReader} implementation for Hbase. */
+public class HBaseSourceSplitReader implements SplitReader<HBaseEvent, HBaseSourceSplit> {
+
+    private final Queue<HBaseSourceSplit> splits;
+    private final HBaseEndpoint hbaseEndpoint;
+
+    @Nullable private String currentSplitId;
+
+    public HBaseSourceSplitReader(byte[] serializedConfig) {
+        try {
+            this.hbaseEndpoint = new HBaseEndpoint(serializedConfig);
+        } catch (Exception e) {
+            throw new RuntimeException("failed HBase consumer", e);
+        }
+        this.splits = new ArrayDeque<>();
+    }
+
+    @Override
+    public RecordsWithSplitIds<HBaseEvent> fetch() throws IOException {
+        final HBaseSourceSplit nextSplit = splits.poll();
+        if (nextSplit != null) {
+            currentSplitId = nextSplit.splitId();
+        }
+        HBaseEvent nextValue = hbaseEndpoint.next();
+        List<HBaseEvent> records = Collections.singletonList(nextValue);
+        return new HbaseSplitRecords<>(currentSplitId, records.iterator(), Collections.emptySet());
+    }
+
+    @Override
+    public void handleSplitsChanges(SplitsChange<HBaseSourceSplit> splitsChanges) {
+        if (splitsChanges instanceof SplitsAddition) {
+            HBaseSourceSplit split = splitsChanges.splits().get(0);
+            try {
+                this.hbaseEndpoint.startReplication(split.getTable(), split.getColumnFamilies());
+            } catch (Exception e) {
+                throw new RuntimeException("failed HBase consumer", e);
+            }
+            splits.addAll(splitsChanges.splits());
+        } else {
+            throw new UnsupportedOperationException(
+                    "Unsupported splits change type "
+                            + splitsChanges.getClass().getSimpleName()
+                            + " in "
+                            + this.getClass().getSimpleName());
+        }
+    }
+
+    @Override
+    public void wakeUp() {}
```

Comment:
This wakeup should complete the future of the `hbaseEndpoint` where you currently call `next`. Thus `next` finishes with empty result and `fetch` can immediately terminate.

Reply:
✅ in 572dc10e8ba6ea81e19116c441e5b3f5794d741f